cli/commands/completion.go

```diff
@@ -0,0 +1,85 @@
+package commands
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/apex/log"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+var (
+	completionsDirName = "completion"
+
+	bashCompFilePath string
+	zshCompFilePath  string
+
+	defaultBashCompFilePath string
+	defaultZshCompFilePath  string
+)
+
+func init() {
+	defaultBashCompFilePath = filepath.Join(completionsDirName, "bash", rootCmd.Name())
+	defaultZshCompFilePath = filepath.Join(completionsDirName, "zsh", fmt.Sprintf("_%s", rootCmd.Name()))
+
+	var genCmd = &cobra.Command{
+		Hidden: true,
+		Use:    "gen",
+		Short:  "Generate completion script",
+		Args:   cobra.MaximumNArgs(0),
+		PreRun: func(cmd *cobra.Command, args []string) {
+			cutFlagsDesc(rootCmd)
+		},
+		Run: func(cmd *cobra.Command, args []string) {
+			err := runGenCmd(cmd, args)
+			if err != nil {
+				log.Fatalf(err.Error())
+			}
+		},
+	}
+
+	rootCmd.AddCommand(genCmd)
+
+	genCmd.Flags().StringVar(&bashCompFilePath, "bash", defaultBashCompFilePath, "Bash completion file path")
+	genCmd.Flags().StringVar(&zshCompFilePath, "zsh", defaultZshCompFilePath, "Zsh completion file path")
+}
+
+// cutFlagsDesc cuts command usage on first '\n'
+// it's needed to make zsh comletion for flags prettier
+func cutFlagsDesc(cmd *cobra.Command) {
+	flags := cmd.Flags()
+	flags.VisitAll(func(f *pflag.Flag) {
+		f.Usage = strings.SplitN(f.Usage, "\n", 2)[0]
```

Comment:
Why "2"? Not "1"?

Reply:
Because 2 is a number of parts, not splits.